feat: adding the annotaTR tool

gymreklab opened the pull request:
This PR introduces the `annotaTR` tool, which can be used to:
* Add dosages to TR VCF files
* Add Beagle annotations to imputed files (previously performed with `trtools_prep_beagle_vcf.sh`)
* Output to either VCF or PGEN format

It required adding pgenlib as a dependency and updating the version of cyvcf2 since we need `cyvcf2.VCF.num_variants`.

This feature is related to and should resolve #209. Steps to solve issues there:

```
# Perform TR imputation using the Saini panel
java -Xmx10g -jar beagle.19Apr22.7c0.jar \
    gt=1kg_snps_NA12878.vcf.gz \
    ref=1kg.snp.str.chr18.vcf.gz impute=true ap=true \
    out=1kg_imputed_chr18
tabix -p vcf 1kg_imputed_chr18.vcf.gz

# Annotate dosages and output both vcf and pgen
annotaTR \
    --vcf 1kg_imputed_chr18.vcf.gz --vcftype hipstr \
    --ref-panel 1kg.snp.str.chr18.vcf.gz \
    --outtype vcf pgen \
    --out test \
    --dosages bestguess_norm 
bgzip test.vcf
tabix -p vcf test.vcf.gz

# Test the pgen output is valid
plink2 --pfile test --validate

# Test the VCF output can be used as input to TRTools
statSTR --vcf test.vcf.gz --vcftype hipstr --afreq --out test
```

## Checklist

* [x ] I've checked to ensure there aren't already other open [pull requests](../../../pulls) for the same update/change
* [x ] I've prefixed the title of my PR according to [the conventional commits specification](https://www.conventionalcommits.org/). If your PR fixes a bug, please prefix the PR with `fix: `. Otherwise, if it introduces a new feature, please prefix it with `feat: `. If it introduces a breaking change, please add an exclamation before the colon, like `feat!: `. If the scope of the PR changes because of a revision to it, please update the PR title, since the title will be used in our CHANGELOG.
* [ x] At the top of the PR, I've [listed any open issues that this PR will resolve](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword). For example, "resolves #0" if this PR resolves issue #0
- [ x] I've explained my changes in a manner that will make it possible for both users and maintainers of TRTools to understand them
* [ x] I've added tests for any new functionality. Or, if this PR fixes a bug, I've added test(s) that replicate it
* [ x] All directories with large test files are listed in [the "exclude" section](https://python-poetry.org/docs/pyproject/#include-and-exclude) of our pyproject.toml so that they do not appear in our PyPI distribution. All new files are also smaller than 0.5 MB.
* [ x] I've updated the relevant REAMDEs with any new usage information and checked that the newly built documentation is formatted properly
* [ x] All functions, modules, classes etc. still conform to [numpy docstring standards](https://numpydoc.readthedocs.io/en/latest/format.html)
* [x ] (if applicable) I've updated the pyproject.toml file with any changes I've made to TRTools's dependencies, and I've run `poetry lock --no-update` to ensure the lock file stays up to date and that our dependencies are locked to their minimum versions
* [ x] In the body of this PR, I've included a short address to the reviewer highlighting one or two items that might deserve their focus

aryarm reviewed (changes requested):
Super exciting work! Thanks for taking this on, @mgymrek 

I took a look and have mostly just minor comments scattered throughout. Honestly, you gave so much attention to small edge cases and validating inputs in this PR, that I ultimately didn't feel like I had much to say, especially for the `annotaTR.py` script. So please let me know if there was anything in particular that you wanted me to look at that I might have missed.

### Code suggestions
- If you haven't already, it might be a good idea to confirm with @LiterallyUniqueLogin and @TaraMirmira that the `bestguess_norm` method will work wrt the clipping? I don't completely understand the clipping, but I think they had specific reasons for the clipping that might have been only relevant when working with Beagle dosages? Or if you're already testing that with @nicholema, then that's also fine.
- It might be a good idea to confirm that the PVAR file can be read by `cyvcf2`, since we will probably want to read from it in downstream tools. Perhaps this can be in a test?
    (Apparently, a VCF without a FORMAT column or any FORMAT fields is still a valid VCF, so the PVAR spec is actually a subset of the VCF spec!)
    Some of the PVAR files seem to have `None` instead of `.` and the INFO fields of each line aren't prefixed by `DSLEN=`
- _Not really a suggestion but just a note for the future_: Downstream users will probably prefer to get information from the PVAR file rather than the VCF, if they can help it. This is because the PVAR file doesn't have sample-specific info, so it will be much faster to read from it. ([See benchmarks for reading TRs from PVAR/PGEN files in haptools.](https://drive.google.com/file/d/1nYENunmcvw6NvA-RHgT1fsv7uCZP2Frx/view?usp=sharing)) Of course, someone could also just create a PVAR file with `plink2` and the VCF output of `annotaTR`, but I'm guessing it will be expensive for annotaTR to write that VCF and then they will also have that extra intermediate step:
    ```
    plink2 --make-just-pvar 'cols=vcfheader,qual,filter,info' --vcf annotaTR.vcf --out output --vcf-half-call m
    ```
    So perhaps we might one day want to create an additional `pvar` output type for annotaTR, so that we can write the equivalent of the VCF output but in a different, separate PVAR file? Unfortunately, I don't think there's an easy way to use `cyvcf2` to write a VCF without FORMAT fields, so I think this idea is dead in the water, but I figured it might be good to mention it anyway. Also, I don't think it's worth holding up this PR over something like that.

### Testing suggestions
The tests are pretty comprehensive for all different kinds of inputs, but I noticed that most of the tests only check the exit codes? It might also be good to check the outputs of annotaTR, as well. I know you already have tests for `GetDosages()`, but somebody might also try to change `annotaTR.py` one day, so we will probably want tests to check that the output of the command-line tool also stays the same. We did something like this in [test_mergeSTR.py](https://github.com/gymrek-lab/TRTools/blob/be63e79bb6402276e82cd7aea85bb76942217f75/trtools/mergeSTR/tests/test_mergeSTR.py#L277-L299), so you could probably copy/adapt that code.

While I was reading through the code, I also wrote down any edge cases I could think of. Here are a few that you might not already have tests for?
1. SNPs and TRs mixed into the file but no `--ref-panel`
2. non indexed VCF file input

And here are the edge cases that it seems like you already have tests for:
1. haploid calls
2. missing and low ploidy calls (-1 and -2)
3. A TR in the input VCF that's not in the ref panel
4. VCF output, PGEN output, both output
5. All of the different kinds of dosages
6. missing AP fields but beagle input requested
7. missing AP fields and no beagle input requested
8. no missing AP fields and no beagle input requested
9. no missing AP fields and beagle input requested
10. duplicate loci to ignore with `--ignore-duplicates`
11. all different kinds of reference panels: hipstr, advntr, gangstr, expansionhunter

gymreklab commented on pyproject.toml:
```diff
@@ -33,7 +33,7 @@ exclude = [
 python = ">=3.7.1,<4.0"
 importlib-metadata = {version = ">=3.0.0", python = "<3.8"}
-cyvcf2 = ">=0.30.4"
+cyvcf2 = ">=0.31.1"
```
ok, thanks for looking into that!

gymreklab commented on trtools/annotaTR/README.rst:
```diff
@@ -0,0 +1,137 @@
+
+* :code:`bestguess`: dosages are computed by summing the length of alleles (given in number of repeat units) for each call. e.g. for a genotype heterozygous for 3 and 4 copies of a repeat, the best guess dosage will be 7.
+* :code:`bestguess_norm`: same as above, but scaled to be between 0 and 2. This is required if generating pgen output, since pgen only supports dosage values in this range.
+* :code:`beagpleap`: dosages for imputed calls are computed in a way that takes into account imputation uncertainty. This option requires the Beagle VCF FORMAT fields :code:`AP1` and :code:`AP2` to be present, which give the probability of each alternate allele on each of the two haplotypes of an individual. AP-based dosages are generated by computing a weighted sum of allele lengths for each hap ...
```
nice!

gymreklab commented on poetry.lock:
thanks for updating those docs! I replaced the `poetry.lock` file with the one you suggested

gymreklab commented on trtools/annotaTR/annotaTR.py:
```diff
@@ -0,0 +1,597 @@
+    if refpanel_metadata is not None:
+        variant_ct = ref_variant_ct
+    else:
+        variant_ct = reader.num_records
```
Yes this requires bgzip/index. This requirement was mentioned in the help message of annotaTR but not in the README so I added it. `checkgz` defaults to True for `utils.LoadSingleReader` so this should be caught but I added `checkgz=True` explicitly to make sure the requirement to have bgzip/index is more clear.

gymreklab commented on trtools/annotaTR/annotaTR.py:
```diff
@@ -0,0 +1,597 @@
+    # variant_ct needed for pgen
+    # If using a ref panel, assume we have same number
+    # of TRs as the panel
+    # Otherwise, assume our file is all TRs and use record count
```
Good point! A mix of TRs and SNPs in the input file with no refpanel specified will result in an error when we try to convert a SNP record to a TRRecord. This would have resulted in a cryptic error before but I now added a check for this and will output a more helpful message before failing now:

```
"Error converting {chrom}:{pos} to a TR record. "
                "If your file is a mix of SNPs/TRs (e.g. from Beagle) you "
                "must provide a reference panel.".format(chrom=record.CHROM, pos=record.POS)
```

gymreklab commented on trtools/utils/utils.py:
```diff
@@ -18,7 +18,9 @@
-        checkgz: bool = True) -> Optional[cyvcf2.VCF]:
+        checkgz: bool = True,
+        lazy: bool = False,
+        samples: List[str] = None) -> Optional[cyvcf2.VCF]:
```
Great suggestion! I have had similar hours for something else spent debugging after assuming sample order stays the same... I changed based on your suggestions.

(Here I just added it to avoid having cyvcf2 avoid loading any samples from the ref panel to save time, since we just need the INFO fields. but definitely this option could lead to problems down the road and set seems more appropriate)

aryarm commented on trtools/utils/tr_harmonizer.py:
```diff
@@ -1065,6 +1083,93 @@ def UniqueStringGenotypes(self) -> Set[int]:
         return set(self.UniqueStringGenotypeMapping().values())
 
+    def GetDosages(self, 
+            dosagetype: TRDosageTypes = TRDosageTypes.bestguess) -> np.ndarray:
```
I just realized that `numpy.typing` requires `numpy>=1.20` and we only have `1.17` in our lock file

Feel free to ignore this suggestion in that case!

gymreklab replied:
ok, will revert back to what we had

gymreklab commented on trtools/utils/tr_harmonizer.py:
```diff
@@ -1065,6 +1083,93 @@ def UniqueStringGenotypes(self) -> Set[int]:
+            # But if normalizing set those to np.nan since unclear
+            # how to normalize haploid dosages which end up negative
+            # under the current method...
+            lengts = self.GetLengthGenotypes()
```
Good point. Modified to return None if no samples, and added a test for this

gymreklab commented on trtools/annotaTR/annotaTR.py:
```diff
@@ -0,0 +1,597 @@
+from trtools import __version__
+
+# Variables for PGEN output
+DEFAULT_PGEN_BATCHSIZE = 1000
```
Good point! Edited to expose as a command line option. Kept default 1000 since our own use cases have been primarily huge files

gymreklab commented on trtools/annotaTR/annotaTR.py:
```diff
@@ -0,0 +1,597 @@
+            return 1
+    if args.vcftype != 'auto':
+        try:
+            checktype = trh.VcfTypes[args.vcftype]
```
correct this was just meant as a check to make sure it is valid input. modified to avoid creating a new variable

gymreklab commented:
Thanks @aryam for the thorough review!

I have addressed the code suggestions and resolved all the conversations. See below regarding the other comments:

## Code suggestions

* Clipping/thresholding is to avoid rounding errors to ensure all the probabilities add up to 1
* Thanks for the suggestion regarding PVAR files as VCFs! I made some minor changes to make them valid (added QUAL/FILTER columns and fileformat header). Also added DSLEN prefix and changed "None" ID's to "."

## Testing suggestions

* I added a command line case for the case of SNPs/TRs mixed in but no ref-panel. This case would have failed since it would attempt to load the SNP as a TR record but now it fails more gracefully: e.g. "Error converting chr21:14245507 to a TR record. If your file is a mix of SNPs/TRs (e.g. from Beagle) you must provide a reference panel."
* Added command line tests for VCF files not being zipped or indexed

## Tests for files remaining the same

I addded tests based on the mergeSTR examples to check if files are unchanged. This required some minor changes to `trtools/testsupport/utils.py`:

* I added an option `max_lines_to_compare` to `assert_same_vcf`. Since I had to just keep the top lines of my example files otherwise they were too big. By default we should still read all lines of files to compare but for my test files I chopped them at 200 lines. (see `trtools/testsupport/sample_vcfs/annotaTR_vcfs/create_test_files.sh`)
* I modified `_make_info_dict` to not crash if an INFO field does not have a value. This is possible for INFO fields with `Number=0` like the `IMP` flag output by Beagle.

aryarm commented on trtools/utils/tr_harmonizer.py:
```diff
@@ -1065,6 +1083,90 @@ def UniqueStringGenotypes(self) -> Set[int]:
+            ref2_dos = ref2*self.ref_allele_length
+
+            # Add together for final dosage
+            unnorm_dosages = (h1_dos + h2_dos + ref1_dos + ref2_dos).astype(np.float32)
```
```suggestion
            unnorm_dosages = (h1_dos + h2_dos + ref1_dos + ref2_dos).astype(np.float32)
        else:
            raise ValueError("Unsupported dosagetype")
```

aryarm reviewed (approved):
ok, sounds good! Thanks for implementing my suggestions

And chopping at 200 lines sounds very reasonable. We shouldn't need more than a few lines, anyway.
